// @quarkly/community-kit/ShareButton/props/overrides.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _fa = require("react-icons/fa");

var _default = {
  Icon: {
    kind: 'Icon',
    props: {
      category: 'fa'
    }
  },
  'Icon Facebook': {
    props: {
      defaultIcon: _fa.FaFacebookF,
      icon: 'FaFacebookF'
    }
  },
  'Icon Twitter': {
    props: {
      defaultIcon: _fa.FaTwitter,
      icon: 'FaTwitter'
    }
  },
  'Icon Telegram': {
    props: {
      defaultIcon: _fa.FaTelegramPlane,
      icon: 'FaTelegramPlane'
    }
  },
  'Icon WhatsApp': {
    props: {
      defaultIcon: _fa.FaWhatsapp,
      icon: 'FaWhatsapp'
    }
  },
  'Icon LinkedIn': {
    props: {
      defaultIcon: _fa.FaLinkedinIn,
      icon: 'FaLinkedinIn'
    }
  },
  'Icon VK': {
    props: {
      defaultIcon: _fa.FaVk,
      icon: 'FaVk'
    }
  },
  'Icon Odnoklassniki': {
    props: {
      defaultIcon: _fa.FaOdnoklassniki,
      icon: 'FaOdnoklassniki'
    }
  },
  'Icon Reddit': {
    props: {
      defaultIcon: _fa.FaRedditAlien,
      icon: 'FaRedditAlien'
    }
  },
  'Icon Tumblr': {
    props: {
      defaultIcon: _fa.FaTumblr,
      icon: 'FaTumblr'
    }
  },
  'Icon Viber': {
    props: {
      defaultIcon: _fa.FaViber,
      icon: 'FaViber'
    }
  },
  'Icon Line': {
    props: {
      defaultIcon: _fa.FaLine,
      icon: 'FaLine'
    }
  },
  'Icon Weibo': {
    props: {
      defaultIcon: _fa.FaWeibo,
      icon: 'FaWeibo'
    }
  },
  'Icon Pocket': {
    props: {
      defaultIcon: _fa.FaGetPocket,
      icon: 'FaGetPocket'
    }
  }
};
exports["default"] = _default;
