// @quarkly/community-kit/DateSinglePicker/DateSinglePicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _DatePicker = _interopRequireDefault(require("../DatePicker"));

var _props = require("./props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var DateSinglePicker = function DateSinglePicker(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      locale = _ref.locale,
      minDate = _ref.minDate,
      maxDate = _ref.maxDate,
      initialMonth = _ref.initialMonth,
      showOutsideDays = _ref.showOutsideDays,
      disabledDates = _ref.disabledDates,
      disabledDaysOfWeek = _ref.disabledDaysOfWeek,
      props = _objectWithoutProperties(_ref, ["value", "onChange", "locale", "minDate", "maxDate", "initialMonth", "showOutsideDays", "disabledDates", "disabledDaysOfWeek"]);

  return /*#__PURE__*/_react["default"].createElement(_DatePicker["default"], _extends({
    locale: locale,
    value: value,
    onChange: onChange,
    minDate: minDate,
    maxDate: maxDate,
    initialMonth: initialMonth,
    showOutsideDays: showOutsideDays,
    disabledDates: disabledDates,
    disabledDaysOfWeek: disabledDaysOfWeek
  }, props, {
    mode: "single",
    numberOfMonths: 1
  }));
};

Object.assign(DateSinglePicker, {
  title: 'DateSinglePicker',
  desciption: {
    en: 'A DateSinglePicker shows a monthly calendar and allows the user to choose a single date.'
  },
  propInfo: _props.propInfo,
  defaultProps: _props.defaultProps
});
var _default = DateSinglePicker;
exports["default"] = _default;
