// @quarkly/community-kit/Slider/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "clamp", {
  enumerable: true,
  get: function get() {
    return _clamp["default"];
  }
});
Object.defineProperty(exports, "approxEqual", {
  enumerable: true,
  get: function get() {
    return _approxEqual["default"];
  }
});
Object.defineProperty(exports, "formatPercentage", {
  enumerable: true,
  get: function get() {
    return _formatPercentage["default"];
  }
});
Object.defineProperty(exports, "countDecimalPlaces", {
  enumerable: true,
  get: function get() {
    return _countDecimalPlaces["default"];
  }
});
Object.defineProperty(exports, "getLabelPrecision", {
  enumerable: true,
  get: function get() {
    return _getLabelPrecision["default"];
  }
});
Object.defineProperty(exports, "formatLabel", {
  enumerable: true,
  get: function get() {
    return _formatLabel["default"];
  }
});
Object.defineProperty(exports, "validateProps", {
  enumerable: true,
  get: function get() {
    return _validateProps["default"];
  }
});
Object.defineProperty(exports, "roundValue", {
  enumerable: true,
  get: function get() {
    return _roundValue["default"];
  }
});

var _clamp = _interopRequireDefault(require("./clamp"));

var _approxEqual = _interopRequireDefault(require("./approxEqual"));

var _formatPercentage = _interopRequireDefault(require("./formatPercentage"));

var _countDecimalPlaces = _interopRequireDefault(require("./countDecimalPlaces"));

var _getLabelPrecision = _interopRequireDefault(require("./getLabelPrecision"));

var _formatLabel = _interopRequireDefault(require("./formatLabel"));

var _validateProps = _interopRequireDefault(require("./validateProps"));

var _roundValue = _interopRequireDefault(require("./roundValue"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
