// @quarkly/community-kit/Slider/utils/validateProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.isNumber = void 0;

var _utils = require("../../utils");

var _props = require("../props");

var getArrayOfNumber = function getArrayOfNumber(value, defaultValue) {
  if (Array.isArray(value)) {
    return value.map(Number);
  }

  if (typeof value === 'string' && value.length > 0) {
    return value.split(',').map(Number);
  }

  return defaultValue;
};

var isNumber = function isNumber(x) {
  return typeof x === 'number';
};

exports.isNumber = isNumber;

var validateProps = function validateProps(_ref) {
  var minFromProps = _ref.minFromProps,
      maxFromProps = _ref.maxFromProps,
      stepSizeFromProps = _ref.stepSizeFromProps,
      labelValuesFromProps = _ref.labelValuesFromProps,
      labelStepSizeFromProps = _ref.labelStepSizeFromProps,
      labelPrecisionFromProps = _ref.labelPrecisionFromProps;
  var min = _props.defaultProps.min,
      max = _props.defaultProps.max,
      stepSize = _props.defaultProps.stepSize,
      labelStepSize = _props.defaultProps.labelStepSize,
      labelPrecision = _props.defaultProps.labelPrecision;
  min = (0, _utils.getNumber)(minFromProps, min);
  max = (0, _utils.getNumber)(maxFromProps, max);
  stepSize = (0, _utils.getNumber)(stepSizeFromProps, stepSize, function (val) {
    return val > 0;
  });
  labelStepSize = (0, _utils.getNumber)(labelStepSizeFromProps, labelStepSize, function (val) {
    return val > 0;
  });
  labelPrecision = (0, _utils.getNumber)(labelPrecisionFromProps, labelPrecision, function (val) {
    return val >= 0;
  });
  var labelValues = getArrayOfNumber(labelValuesFromProps, undefined);
  return {
    min: min,
    max: max,
    stepSize: stepSize,
    labelStepSize: labelStepSize,
    labelPrecision: labelPrecision,
    labelValues: labelValues
  };
};

var _default = validateProps;
exports["default"] = _default;
