// @quarkly/community-kit/modules/knobs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _addonKnobs = require("@storybook/addon-knobs");

var _default = function _default() {
  var defaults = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    // Layout
    display: (0, _addonKnobs.select)('display', {
      block: 'block',
      inline: 'inline',
      'inline-block': 'inline-block',
      flex: 'flex',
      grid: 'grid',
      none: 'none',
      contents: 'contents',
      'inline-flex': 'inline-flex',
      'inline-grid': 'inline-grid',
      'inline-table': 'inline-table',
      'list-item': 'list-item',
      'run-in': 'run-in',
      inherit: 'inherit',
      initial: 'initial'
    }, defaults.display || 'inline-block', 'Layout'),
    width: (0, _addonKnobs.text)('width', defaults.width || '', 'Layout'),
    'min-width': (0, _addonKnobs.text)('minWidth', defaults['min-width'] || '', 'Layout'),
    'max-width': (0, _addonKnobs.text)('maxWidth', defaults['max-width'] || '', 'Layout'),
    height: (0, _addonKnobs.text)('height', defaults.height || '', 'Layout'),
    'min-height': (0, _addonKnobs.text)('minHeight', defaults['min-height'] || '', 'Layout'),
    'max-height': (0, _addonKnobs.text)('maxHeight', defaults['max-height'] || '', 'Layout'),
    'overflow-x': (0, _addonKnobs.select)('overflowX', {
      '–': 'initial',
      auto: 'auto',
      hidden: 'hidden',
      scroll: 'scroll',
      visible: 'visible',
      inherit: 'inherit'
    }, defaults['overflow-x'] || 'visible', 'Layout'),
    'overflow-y': (0, _addonKnobs.select)('overflowY', {
      '–': 'initial',
      auto: 'auto',
      hidden: 'hidden',
      scroll: 'scroll',
      visible: 'visible',
      inherit: 'inherit'
    }, defaults['overflow-y'] || 'visible', 'Layout'),
    'flex-direction': (0, _addonKnobs.select)('flexDirection', {
      '–': 'initial',
      row: 'row',
      'row-reverse': 'row-reverse',
      column: 'column',
      'column-reverse': 'column-reverse'
    }, defaults['flex-direction'] || 'row', 'Layout'),
    'align-items': (0, _addonKnobs.select)('alignItems', {
      '–': 'initial',
      'flex-start': 'flex-start',
      'flex-end': 'flex-end',
      center: 'center',
      baseline: 'baseline',
      stretch: 'stretch',
      inherit: 'inherit',
      initial: 'initial',
      unset: 'unset'
    }, defaults['align-items'] || '–', 'Layout'),
    'justify-content': (0, _addonKnobs.select)('justifyContent', {
      '–': 'initial',
      center: 'center',
      start: 'start',
      end: 'end',
      'flex-start': 'flex-start',
      'flex-end': 'flex-end',
      left: 'left',
      right: 'right',
      baseline: 'baseline',
      'first baseline': 'first baseline',
      'last baseline': 'last baseline',
      'space-between': 'space-between',
      'space-around': 'space-around',
      'space-evenly': 'space-evenly',
      stretch: 'stretch',
      'safe center': 'safe center',
      'unsafe center': 'unsafe center',
      inherit: 'inherit',
      initial: 'initial',
      unset: 'unset'
    }, '–', defaults['justify-content'] || 'Layout'),
    'flex-wrap': (0, _addonKnobs.select)('flexWrap', {
      '–': 'initial',
      nowrap: 'nowrap',
      wrap: 'wrap',
      'wrap-reverse': 'wrap-reverse'
    }, defaults['flex-wrap'] || 'nowrap', 'Layout'),
    // Typography
    font: (0, _addonKnobs.text)('font', defaults.font || '', 'Typography'),
    'font-family': (0, _addonKnobs.text)('fontFamily', defaults['font-family'] || '', 'Typography'),
    'font-size': (0, _addonKnobs.text)('fontSize', defaults['font-size'] || '', 'Typography'),
    'line-height': (0, _addonKnobs.text)('lineHeight', defaults['line-height'] || '', 'Typography'),
    'font-weight': (0, _addonKnobs.text)('fontWeight', defaults['font-weight'] || '', 'Typography'),
    'font-style': (0, _addonKnobs.select)('fontStyle', {
      '–': 'initial',
      normal: 'normal',
      italic: 'italic'
    }, defaults['font-style'] || '–', 'Typography'),
    'font-variant': (0, _addonKnobs.select)('fontVariant', {
      '–': 'initial',
      normal: 'normal',
      'small-caps': 'small-caps'
    }, defaults['font-variant'] || '–', 'Typography'),
    'letter-spacing': (0, _addonKnobs.text)('letterSpacing', defaults['letter-spacing'] || '', 'Typography'),
    'white-space': (0, _addonKnobs.select)('whiteSpace', {
      '–': 'initial',
      normal: 'normal',
      nowrap: 'nowrap',
      pre: 'pre',
      'pre-line': 'pre-line',
      'pre-wrap': 'pre-wrap'
    }, defaults['white-space'] || '–', 'Typography'),
    'text-transform': (0, _addonKnobs.text)('textTransform', defaults['text-transform'] || '', 'Typography'),
    'text-decoration': (0, _addonKnobs.text)('textDecoration', defaults['text-decoration'] || '', 'Typography'),
    'text-align': (0, _addonKnobs.select)('textAlign', {
      left: 'left',
      center: 'center',
      right: 'right',
      justify: 'justify'
    }, defaults['text-align'] || 'left', 'Typography'),
    color: (0, _addonKnobs.color)('color', defaults.color || '', 'Typography'),
    // Indents
    'margin-top': (0, _addonKnobs.text)('marginTop', defaults['margin-top'] || '', 'Indents'),
    'margin-right': (0, _addonKnobs.text)('marginRight', defaults['margin-right'] || '', 'Indents'),
    'margin-bottom': (0, _addonKnobs.text)('marginBottom', defaults['margin-bottom'] || '', 'Indents'),
    'margin-left': (0, _addonKnobs.text)('marginLeft', defaults['margin-left'] || '', 'Indents'),
    'padding-top': (0, _addonKnobs.text)('paddingTop', defaults['padding-top'] || '', 'Indents'),
    'padding-right': (0, _addonKnobs.text)('paddingRight', defaults['padding-right'] || '', 'Indents'),
    'padding-bottom': (0, _addonKnobs.text)('paddingBottom', defaults['padding-bottom'] || '', 'Indents'),
    'padding-left': (0, _addonKnobs.text)('paddingLeft', defaults['padding-left'] || '', 'Indents'),
    // Background
    background: (0, _addonKnobs.text)('background', defaults.background || '', 'Background'),
    'background-image': (0, _addonKnobs.text)('backgroundImage', defaults['background-image'] || 'url()', 'Background'),
    'background-position': (0, _addonKnobs.text)('backgroundPosition', defaults['background-position'] || 'none', 'Background'),
    'background-size': (0, _addonKnobs.select)('backgroundSize', {
      none: 'none',
      cover: 'cover',
      contain: 'contain'
    }, defaults['background-size'] || 'none', 'Background'),
    'background-repeat': (0, _addonKnobs.select)('backgroundRepeat', {
      '–': 'initial',
      repeat: 'repeat',
      'repeat-x': 'repeat-x',
      'repeat-y': 'repeat-y',
      'no-repeat': 'no-repeat',
      space: 'space',
      round: 'round'
    }, defaults['background-repeat'] || '–', 'Background'),
    'background-attachment': (0, _addonKnobs.select)('backgroundAttachment', {
      '–': 'initial',
      fixed: 'fixed',
      local: 'local',
      scroll: 'scroll'
    }, defaults['background-attachment'] || '–', 'Background'),
    'background-clip': (0, _addonKnobs.select)('backgroundClip', {
      '–': 'initial',
      'border-box': 'border-box',
      'padding-box': 'padding-box',
      'content-box': 'content-box'
    }, defaults['background-clip'] || '–', 'Background'),
    // Border
    'border-width': (0, _addonKnobs.text)('borderWidth', defaults['border-width'] || '0px', 'Border'),
    'border-style': (0, _addonKnobs.select)('borderStyle', {
      none: 'none',
      dotted: 'dotted',
      dashed: 'dashed',
      solid: 'solid',
      "double": 'double',
      groove: 'groove',
      ridge: 'ridge',
      inset: 'inset',
      outset: 'outset',
      hidden: 'hidden',
      inherit: 'inherit'
    }, defaults['border-style'] || 'none', 'Border'),
    'border-color': (0, _addonKnobs.color)('borderColor', defaults['border-color'] || 'rgb(118,118,118)', 'Border'),
    'border-radius': (0, _addonKnobs.text)('borderRadius', defaults['border-radius'] || '', 'Border'),
    // Position
    position: (0, _addonKnobs.select)('position', {
      '–': 'initial',
      absolute: 'absolute',
      fixed: 'fixed',
      sticky: 'sticky',
      relative: 'relative',
      "static": 'static',
      inherit: 'inherit'
    }, defaults.position || '–', 'Position'),
    top: (0, _addonKnobs.text)('top', defaults.top || '', 'Position'),
    right: (0, _addonKnobs.text)('right', defaults.right || '', 'Position'),
    bottom: (0, _addonKnobs.text)('bottom', defaults.bottom || '', 'Position'),
    left: (0, _addonKnobs.text)('left', defaults.left || '', 'Position'),
    'z-index': (0, _addonKnobs.text)('zIndex', defaults['z-index'] || 'auto', 'Position'),
    // Effects
    opacity: (0, _addonKnobs.number)('opacity', defaults.opacity || '1', '{1}', 'Effects'),
    cursor: (0, _addonKnobs.select)('cursor', {
      '–': 'initial',
      auto: 'auto',
      "default": 'default',
      none: 'none',
      'context-menu': 'context-menu',
      help: 'help',
      pointer: 'pointer',
      progress: 'progress',
      wait: 'wait',
      cell: 'cell',
      crosshair: 'crosshair',
      text: 'text',
      'vertical-text': 'vertical-text',
      alias: 'alias',
      copy: 'copy',
      move: 'move',
      'no-drop': 'no-drop',
      'not-allowed': 'not-allowed',
      'e-resize': 'e-resize',
      'n-resize': 'n-resize',
      'ne-resize': 'ne-resize',
      'nw-resize': 'nw-resize',
      's-resize': 's-resize',
      'se-resize': 'se-resize',
      'sw-resize': 'sw-resize',
      'w-resize': 'w-resize',
      'ew-resize': 'ew-resize',
      'ns-resize': 'ns-resize',
      'nesw-resize': 'nesw-resize',
      'nwse-resize': 'nwse-resize',
      'col-resize': 'col-resize',
      'row-resize': 'row-resize',
      'all-scroll': 'all-scroll',
      'zoom-in': 'zoom-in',
      'zoom-out': 'zoom-out',
      grab: 'grab',
      grabbing: 'grabbing'
    }, defaults.cursor || '–', 'Effects'),
    'blend-mode': (0, _addonKnobs.select)('blendMode', {
      '–': 'initial',
      normal: 'normal',
      multiply: 'multiply',
      screen: 'screen',
      overlay: 'overlay',
      darken: 'darken',
      lighten: 'lighten',
      'color-dodge': 'color-dodge',
      'color-burn': 'color-burn',
      'hard-light': 'hard-light',
      'soft-light': 'soft-light',
      difference: 'difference',
      exclusion: 'exclusion',
      hue: 'hue',
      saturation: 'saturation',
      color: 'color',
      luminosity: 'luminosity'
    }, defaults['blend-mode'] || '–', 'Effects'),
    'box-shadow': (0, _addonKnobs.text)('boxShadow', defaults['box-shadow'] || 'none', 'Effects'),
    'text-shadow': (0, _addonKnobs.text)('textShadow', defaults['text-shadow'] || 'none', 'Effects'),
    transition: (0, _addonKnobs.text)('transition', defaults.transition || 'all 0s ease 0s', 'Effects'),
    transform: (0, _addonKnobs.text)('transform', defaults.transform || 'none', 'Effects'),
    filter: (0, _addonKnobs.text)('filter', defaults.filter || 'none', 'Effects')
  };
};

exports["default"] = _default;
