// @quarkly/community-kit/AccordionItemDetails.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@quarkly/components");

var _widgets = require("@quarkly/widgets");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var overrides = {
  'Accordion Details': {
    kind: 'Box',
    props: {
      'margin-bottom': '-1px',
      'max-height': '0px'
    }
  },
  'Accordion Details :open': {
    kind: 'Box',
    props: {
      'margin-bottom': '0px',
      'min-height': '16px',
      'max-height': 'auto'
    }
  },
  'Accordion Details :disabled': {
    kind: 'Box',
    props: {}
  }
};

var AccordionItemDetails = function AccordionItemDetails(_ref) {
  var open = _ref.open,
      disabled = _ref.disabled,
      props = _objectWithoutProperties(_ref, ["open", "disabled"]);

  var _useOverrides = (0, _components.useOverrides)(props, overrides, {}),
      override = _useOverrides.override,
      children = _useOverrides.children,
      rest = _useOverrides.rest;

  return /*#__PURE__*/_react["default"].createElement(_widgets.Box, _extends({
    padding: "0 24px",
    "min-height": "0px",
    "border-bottom": "1px solid #BEC7CC",
    "box-sizing": "border-box",
    overflow: "hidden"
  }, rest, override('Accordion Details', open && 'Accordion Details :open', disabled && 'Accordion Details :disabled')), children);
};

var _default = Object.assign(AccordionItemDetails, {
  overrides: overrides
});

exports["default"] = _default;
