// @quarkly/community-kit/FbComment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _widgets = require("@quarkly/widgets");

var _reactFacebook = require("react-facebook");

var _ComponentNotice = _interopRequireDefault(require("./ComponentNotice"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var FacebookComments = function FacebookComments(_ref) {
  var appId = _ref.appId,
      language = _ref.language,
      href = _ref.href,
      rest = _objectWithoutProperties(_ref, ["appId", "language", "href"]);

  return /*#__PURE__*/_react["default"].createElement(_widgets.Box, rest, appId && href ? /*#__PURE__*/_react["default"].createElement(_reactFacebook.FacebookProvider, {
    appId: appId,
    key: appId + language,
    language: language
  }, /*#__PURE__*/_react["default"].createElement(_reactFacebook.Comments, {
    href: href
  })) : /*#__PURE__*/_react["default"].createElement(_ComponentNotice["default"], {
    message: "Add your Facebook App ID and 'Link to coments' in the Props panel"
  }));
};

var propInfo = {
  appId: {
    title: {
      en: 'Facebook app ID',
      ru: 'ID приложения Facebook'
    },
    control: 'input',
    type: 'text',
    category: 'Main',
    weight: 1
  },
  href: {
    title: {
      en: 'Link to comments',
      ru: 'Ссылка на комментарии'
    },
    control: 'input',
    type: 'text',
    category: 'Main',
    weight: 1
  },
  language: {
    title: {
      en: 'The language of the widget',
      ru: 'Язык загружаемого виджета'
    },
    control: 'select',
    variants: [{
      title: 'English',
      value: 'en_US'
    }, {
      title: 'Русский',
      value: 'ru_RU'
    }],
    category: 'Main',
    weight: 1
  }
};
var defaultProps = {
  language: 'en_US'
};
Object.assign(FacebookComments, {
  title: 'Facebook Comments',
  description: {
    en: 'This component allows you to add a comment form for Facebook',
    ru: 'Компонент для добавления формы комментариев Facebook'
  },
  propInfo: propInfo,
  defaultProps: defaultProps
});
var _default = FacebookComments;
exports["default"] = _default;
