// @quarkly/community-kit/DatePicker/DatePicker.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _widgets = require("@quarkly/widgets");

var _components = require("@quarkly/components");

var _DatePicker = require("./contexts/DatePicker");

var _Override = require("./contexts/Override");

var _components2 = require("./components");

var _utils = require("./utils");

var _props = require("./props");

var _SelectSingle = require("./contexts/SelectSingle");

var _SelectRange = require("./contexts/SelectRange");

var _utils2 = require("../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var DatePicker = function DatePicker(_ref) {
  var mode = _ref.mode,
      value = _ref.value,
      onChange = _ref.onChange,
      showOutsideDays = _ref.showOutsideDays,
      numberOfMonthsFromProps = _ref.numberOfMonths,
      localeFromProps = _ref.locale,
      minDateFromProps = _ref.minDate,
      maxDateFromProps = _ref.maxDate,
      initialMonthFromProps = _ref.initialMonth,
      disabledDatesFromProps = _ref.disabledDates,
      disabledDaysOfWeekFromProps = _ref.disabledDaysOfWeek,
      props = _objectWithoutProperties(_ref, ["mode", "value", "onChange", "showOutsideDays", "numberOfMonths", "locale", "minDate", "maxDate", "initialMonth", "disabledDates", "disabledDaysOfWeek"]);

  var _useState = (0, _react.useState)(_utils.locales.enUS),
      _useState2 = _slicedToArray(_useState, 2),
      locale = _useState2[0],
      setLocale = _useState2[1];

  var _useOverrides = (0, _components.useOverrides)(props, _props.overrides),
      override = _useOverrides.override,
      rest = _useOverrides.rest;

  (0, _react.useEffect)(function () {
    if (_utils.locales[localeFromProps]) setLocale(_utils.locales[localeFromProps]);
  }, [localeFromProps]);
  var minDate = (0, _react.useMemo)(function () {
    return (0, _utils.dateParse)(minDateFromProps === null || minDateFromProps === void 0 ? void 0 : minDateFromProps.toString());
  }, [minDateFromProps]);
  var maxDate = (0, _react.useMemo)(function () {
    return (0, _utils.dateParse)(maxDateFromProps === null || maxDateFromProps === void 0 ? void 0 : maxDateFromProps.toString());
  }, [maxDateFromProps]);
  var numberOfMonths = (0, _react.useMemo)(function () {
    var parsedValue = Number.parseInt(numberOfMonthsFromProps, 10);
    if (Number.isNaN(parsedValue) || parsedValue <= 0) return 1;
    return parsedValue;
  }, [numberOfMonthsFromProps]);
  var initialMonth = (0, _react.useMemo)(function () {
    return (0, _utils.dateParse)(initialMonthFromProps === null || initialMonthFromProps === void 0 ? void 0 : initialMonthFromProps.toString(), true);
  }, [initialMonthFromProps]);
  var disabledDates = (0, _react.useMemo)(function () {
    if (!(0, _utils2.isString)(disabledDatesFromProps) || disabledDatesFromProps.length === 0) return [];
    return disabledDatesFromProps.split(',').map(function (x) {
      return (0, _utils.dateParse)(x);
    }).filter(function (x) {
      return x !== undefined;
    });
  }, [disabledDatesFromProps]);
  var disabledDaysOfWeek = (0, _react.useMemo)(function () {
    if (!(0, _utils2.isString)(disabledDaysOfWeekFromProps) || disabledDaysOfWeekFromProps.length === 0) return [];
    return disabledDaysOfWeekFromProps.split(',').map(function (x) {
      return Number.parseInt(x, 10);
    }).filter(function (x) {
      return x >= 0 && x <= 6;
    });
  }, [disabledDaysOfWeekFromProps]);
  var providerProps = {
    mode: mode,
    value: value,
    onChange: onChange,
    locale: locale,
    minDate: minDate,
    maxDate: maxDate,
    initialMonth: initialMonth,
    disabledDates: disabledDates,
    disabledDaysOfWeek: disabledDaysOfWeek,
    numberOfMonths: numberOfMonths,
    showOutsideDays: showOutsideDays
  };
  var SelectProvider = (0, _react.useMemo)(function () {
    var _providers$mode;

    var providers = {
      single: _SelectSingle.SelectSingleProvider,
      range: _SelectRange.SelectRangeProvider
    };
    return (_providers$mode = providers[mode]) !== null && _providers$mode !== void 0 ? _providers$mode : providers.single;
  }, [mode]);
  var selectProps = {
    value: value,
    onChange: onChange
  };
  return /*#__PURE__*/_react["default"].createElement(_widgets.Box, _extends({
    "aria-label": locale.labels.calendar,
    role: "application",
    display: "inline-flex",
    "flex-direction": "column",
    background: "white",
    p: 6
  }, rest), /*#__PURE__*/_react["default"].createElement(_Override.OverrideProvider, {
    value: override
  }, /*#__PURE__*/_react["default"].createElement(_DatePicker.DatePickerProvider, providerProps, /*#__PURE__*/_react["default"].createElement(SelectProvider, selectProps, /*#__PURE__*/_react["default"].createElement(_components2.Calendar, null)))));
};

Object.assign(DatePicker, {
  title: 'DatePicker',
  desciption: {
    en: 'A DatePicker shows a monthly calendar and allows the user to choose a single date.'
  },
  propInfo: _props.propInfo,
  defaultProps: _props.defaultProps
});
var _default = DatePicker;
exports["default"] = _default;
