// @quarkly/community-kit/TabsFullBody/TabsFullBody.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _widgets = require("@quarkly/widgets");

var _components = require("@quarkly/components");

var _TabsFull = require("../TabsFull");

var _ComponentNotice = _interopRequireDefault(require("../ComponentNotice"));

var _utils = require("../utils");

var _overrides = _interopRequireDefault(require("./props/overrides"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var TabsFullBody = function TabsFullBody(props) {
  var _useOverrides = (0, _components.useOverrides)(props, _overrides["default"]),
      override = _useOverrides.override,
      children = _useOverrides.children,
      rest = _useOverrides.rest;

  var context = (0, _TabsFull.useTabs)();
  return /*#__PURE__*/_react["default"].createElement(_widgets.Box, rest, context ? /*#__PURE__*/_react["default"].createElement(_widgets.Box, override('Wrapper'), children, (0, _utils.isEmptyChildren)(children) && /*#__PURE__*/_react["default"].createElement(_widgets.Placeholder, {
    message: "Drop TabsFullContent here"
  })) : /*#__PURE__*/_react["default"].createElement(_ComponentNotice["default"], {
    message: "Place this component inside TabsFull"
  }));
};

Object.assign(TabsFullBody, {
  title: 'TabsFullBody',
  description: {
    en: 'Container for TabsFullContent. This component must be located inside TabsFull',
    ru: 'Контейнер для TabsFullContent. Этот компонент должен располагаться внутри TabsFull'
  },
  overrides: _overrides["default"]
});
var _default = TabsFullBody;
exports["default"] = _default;
