// @quarkly/community-kit/ShareButton/utils/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _isMobile = _interopRequireDefault(require("./isMobile"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var q = function q(obj) {
  Object.keys(obj).forEach(function (key) {
    return obj[key] === undefined && delete obj[key];
  });
  return "?".concat(new URLSearchParams(obj).toString());
};

var _default = {
  Facebook: {
    url: 'https://www.facebook.com/sharer/sharer.php',
    color: '#1778f2',
    query: function query(_ref) {
      var url = _ref.url,
          quote = _ref.quote,
          hashtag = _ref.hashtag;
      return q({
        u: url,
        quote: quote,
        hashtag: hashtag
      });
    }
  },
  Twitter: {
    url: 'https://twitter.com/share',
    color: '#00aced',
    query: function query(_ref2) {
      var url = _ref2.url,
          quote = _ref2.quote;
      return q({
        url: url,
        text: quote
      });
    }
  },
  Telegram: {
    url: 'https://telegram.me/share/url',
    color: '#37aee2',
    query: function query(_ref3) {
      var url = _ref3.url,
          quote = _ref3.quote;
      return q({
        url: url,
        text: quote
      });
    }
  },
  WhatsApp: {
    url: "https://".concat((0, _isMobile["default"])() ? 'api' : 'web', ".whatsapp.com/send"),
    color: '#25D366',
    query: function query(_ref4) {
      var url = _ref4.url,
          quote = _ref4.quote;
      return q({
        text: "".concat(quote, " ").concat(url)
      });
    }
  },
  LinkedIn: {
    url: 'https://linkedin.com/shareArticle',
    color: '#007fb1',
    query: function query(_ref5) {
      var url = _ref5.url,
          title = _ref5.title,
          description = _ref5.description;
      return q({
        mini: 'true',
        url: url,
        title: title,
        summary: description
      });
    }
  },
  VK: {
    url: 'https://vk.com/share.php',
    color: '#2787F5',
    query: function query(_ref6) {
      var url = _ref6.url,
          title = _ref6.title;
      return q({
        url: url,
        title: title
      });
    }
  },
  Odnoklassniki: {
    url: 'https://connect.ok.ru/offer',
    color: '#f2720c',
    query: function query(_ref7) {
      var url = _ref7.url,
          title = _ref7.title,
          description = _ref7.description,
          image = _ref7.image;
      return q({
        url: url,
        title: title,
        description: description,
        imageUrl: image
      });
    }
  },
  Reddit: {
    url: 'https://www.reddit.com/submit',
    color: '#ff4500',
    query: function query(_ref8) {
      var url = _ref8.url,
          title = _ref8.title;
      return q({
        url: url,
        title: title
      });
    }
  },
  Tumblr: {
    url: 'https://www.tumblr.com/widgets/share/tool',
    color: '#2c4762',
    query: function query(_ref9) {
      var url = _ref9.url,
          title = _ref9.title,
          description = _ref9.description;
      return q({
        canonicalUrl: url,
        title: title,
        caption: description,
        posttype: 'link'
      });
    }
  },
  Viber: {
    url: 'viber://forward',
    color: '#7C529E',
    query: function query(_ref10) {
      var url = _ref10.url,
          quote = _ref10.quote;
      return q({
        text: "".concat(quote, " ").concat(url)
      });
    }
  },
  Line: {
    url: 'https://social-plugins.line.me/lineit/share',
    color: '#00b800',
    query: function query(_ref11) {
      var url = _ref11.url,
          title = _ref11.title;
      return q({
        url: url,
        text: title
      });
    }
  },
  Weibo: {
    url: 'http://service.weibo.com/share/share.php',
    color: '#CD201F',
    query: function query(_ref12) {
      var url = _ref12.url,
          title = _ref12.title,
          image = _ref12.image;
      return q({
        url: url,
        title: title,
        pic: image
      });
    }
  },
  Pocket: {
    url: 'https://getpocket.com/save',
    color: '#EF3F56',
    query: function query(_ref13) {
      var url = _ref13.url,
          title = _ref13.title;
      return q({
        url: url,
        title: title
      });
    }
  }
};
exports["default"] = _default;
